---
title: "Script 3: MethylAid and PCA"
output:
  html_document:
    keep_md: false
    smart: false
    toc: true
    toc_float: true
    theme: united
date: 'Compiled: `r format(Sys.Date(), "%B %d, %Y")`'
---

```{r, setup, include=FALSE}
knitr::opts_knit$set(
  root.dir = "/exports/molepi/users/ljsinke/GOTO/main/GOTO_Data/IDAT_files/")
```

**This script removes outliers identified by MethylAid and single (unpaired) samples, reruns MethylAid to confirm sample quality, inspects beta value distributions, and performs PCA.**

***

# Setup

Load packages

```{r message=FALSE, warning=FALSE}
library(tidyverse)
library(DNAmArray)
library(minfi)
library(omicsPrint)
library(irlba)
library(BiocParallel)
library(FDb.InfiniumMethylation.hg19)
library(ggfortify)
library(GenomicRanges)
library(SummarizedExperiment)
library(MethylAid)
library(ggrepel)
```

***

# Annotation

We use `DNAmArray` internal files, which were pulled recently (2022) from Zhou's github: https://zwdzwd.github.io/InfiniumAnnotation

```{r warning=FALSE, message = FALSE}
anno <- read_tsv(
  "../../../../LLS/Shared_Data/Manifests/EPIC.hg19.manifest.tsv.gz")
```

We are interested in the:

* `cpg` - ID of the probe for the CpG
* `chr` - Chromosome where the CpG is located
* `start` - Start position of the CpG
* `end` - End position of the CpG
* `strand` - Strand of the CpG
* `gene_HGNC` - Annotated gene 

```{r}
anno <- anno %>% 
  dplyr::select(
    cpg = probeID,
    chr = CpG_chrm,
    start = CpG_beg,
    end = CpG_end,
    strand = probe_strand,
    gene_HGNC,
    MASK_general
  ) %>% 
  mutate(
    chr = substr(chr,4,5)
  )
```

***

# Targets

Load in targets file from wave 1

```{r}
load("../Processing/GOTO_targets-unfiltered.Rdata")
```

Load in MethylAid outliers

```{r}
outliers <- read_csv("../Processing/MethylAid/GOTO_MethylAid-Outliers.csv")
```

Remove outliers and their pairs from the data

```{r}
outlier_pair <- (targets_goto %>% 
  filter(Basename %in% outliers$ID) %>% 
  mutate(
    pair = paste0(IOP2_ID, tissue)
  ))$pair

targets_wave1 <- targets_goto %>% 
  mutate(
    pair = paste0(IOP2_ID, tissue)
  ) %>% 
  filter(
    !pair %in% outlier_pair) %>% 
  dplyr::select(-pair)

print(paste0("Data on ", nrow(targets_wave1), " samples remains..."))
```

Add wave 2 targets

```{r}
load("../Sample_Sheets/GOTO_wave2-targets.Rdata")

targets_wave2 <- targets %>% 
  mutate(
    DNA_labnr = as.factor(DNA_labnr),
    IOP2_ID = as.factor(IOP2_ID),
    HMU_ID = as.factor(HMU_ID),
    tissue = as.factor(tissue),
    timepoint = factor(timepoint, 
                       levels = c("before", "after")),
    sex = factor(sex, 
                 levels=c("male", "female")),
    op_status = factor(op_status, 
                       levels = c("partner", "offspring")),
    plate = factor(plate),
    well = as.factor(paste0(rep(LETTERS[1:8], 3), rep(c("09", "10", "11"), each=8))),
    Basename = paste0(chip_n, "_", chip_position),
    array_n = factor(chip_n),
    array_row = as.numeric(substr(chip_position,3,3))
  ) %>% 
  dplyr::select(
    DNA_labnr, IOP2_ID, HMU_ID, 
    tissue, timepoint, sex, 
    age, bmi, op_status,
    plate, well, isolationdate, 
    conc_ngul, A260280, volume, 
    array_n, array_row, Basename)
```

Combine waves

```{r}
targets <- rbind(targets_wave1, targets_wave2)
```

Make pair ID

```{r}
targets <- targets %>% 
  mutate(
    pair = paste0(IOP2_ID, tissue)
  ) 
```

Remove non-paired samples and one sample where pair was resent in wave 2 (fasted blood_2302)

```{r}
unpaired <- targets %>% 
  group_by(pair) %>% 
  dplyr::summarize(n=n()) %>% 
  ungroup()

unpaired <- (unpaired %>% 
  filter(n<2))$pair

targets <- targets %>% filter(
  (!pair %in% unpaired)) %>% 
  dplyr::select(
    -pair
  )

print(paste0("Data on ", nrow(targets), " samples remains..."))
```

***

# MethylAid

Set `BPPARAM`

```{r}
BPPARAM <- MulticoreParam(8)
```

Summarize IDAT files for MethylAid

```{r message=FALSE, warning=FALSE}
sData_goto <- MethylAid::summarize(
  targets, 
  batchSize=50, 
  BPPARAM = BPPARAM,
  force=TRUE)
```

Save

```{r}
save(
  sData_goto, 
  file="../Processing/MethylAid/GOTO_sData-all.Rdata")
```

This data was visualized using MethylAid and no outliers were identified

***

# Read in Data

Read IDATs into R as an `RGSet` object

```{r warning=FALSE, message=FALSE}
register(MulticoreParam(6))

RGset <- read.metharray.exp(
  targets, 
  base = ".",
  extended=TRUE)
RGset
```

Calculate beta values from the red and green intensities

```{r}
betas <- getBeta(
  RGset, 
  type="Illumina")

dim(betas)
```

Look at the beta value distribution

```{r}
densityPlot(
  RGset, 
  main="Beta density plot", 
  xlab="Beta values")
```

***

# PCA

Calculate PCs

```{r}
pc <- prcomp_irlba(
  t(betas), 
  n=6)

summary(pc)
```

Add PCs to targets

```{r}
targets <- cbind(
  targets, 
  pc$x)
```

The PCA analysis can result in two different symmetrical clusters. This code handles those two eventualities and saves IDs to remove.

```{r}
if(nrow(targets %>% 
        dplyr::filter(PC1 < -30 & tissue != "fasted blood")) == 2) {
  pc_blood <- "right"
} else {
  pc_blood <- "left"
}
```

Plot the first two PCs against each other to ensure the primary source of variability is tissue.

```{r warning=F}
if(pc_blood == "right"){
  pca_plot <- targets %>% 
  ggplot(aes(
    x=PC1,
    y=PC2,
    label = ifelse(
        (PC1 < -30 & tissue != "fasted blood") |
        (PC1 > -60 & tissue == "fasted blood"), 
        as.character(IOP2_ID), NA)
  )) 
} else {
  pca_plot <- targets %>% 
  ggplot(aes(
    x=PC1,
    y=PC2,
    label = ifelse(
        (PC1 > 30 & tissue != "fasted blood") |
        (PC1 < 60 & tissue == "fasted blood"), 
        as.character(IOP2_ID), NA)
  )) 
}

pca_plot +
  geom_point(
    aes(
      colour = tissue)) +
  geom_label_repel(
    aes(color = tissue),
    fill = "white"
  ) +
  xlab("PC1") +
  ylab("PC2") +
  ggtitle("PCA plot coloured by tissue") +
  theme(
    axis.text = element_text(
       size=9, 
       color="#1B2021"),
    axis.title = element_text(
      size=12, 
      hjust=0.5, 
      color="#1B2021"),
    plot.title = element_text(
      size=16, 
      hjust=0.5,
      face="bold", 
      color = "#548687"),
    panel.background = element_rect(
      fill="white", color="#1B2021"),
    panel.grid.major = element_blank(),
    panel.grid.minor = element_blank(),
    plot.background = element_rect(
      fill="white"))
```

There are a few outliers than should be removed:

* One fat sample and its pair, which clusters closely to the blood samples. It is likely this was a very bloody biopsy.
* One muscle sample and its pair, which similarly clusters close to the blood samples
* Two blood samples and their pairs, which cluster close to the fat samples. Potentially, these individual had higher levels of lipids in their blood.

Save the outliers

```{r}
outliers <- targets %>% 
  dplyr::filter(
      (IOP2_ID == "61479" & tissue == "fasted blood") |
      (IOP2_ID == "60871" & tissue == "fasted blood") |
      (IOP2_ID == "61093" & tissue == "fat") |
      (IOP2_ID == "60822" & tissue == "muscle"))
```

Remove them from targets

```{r}
targets <- targets[!(targets$Basename %in% outliers$Basename),]
```

***

# Add smoking

Import data

```{r}
smoke_df <- read_csv('../Pheno/GOTO_SmokingData_20201117_v2.csv')
```

Create smoking var and inspect

```{r}
smoke_df <- smoke_df %>% 
  mutate(smoke = case_when(
    (CurrentSmoking == 1) ~ 2,
    (CurrentSmoking == 0) & (EverSmoked == 1) ~ 1,
    (CurrentSmoking == 0) & (EverSmoked == 0) ~ 0
  ))
xtabs(~smoke_df$smoke + smoke_df$CurrentSmoking)
xtabs(~smoke_df$smoke + smoke_df$EverSmoked)
```

Remove original vars

```{r}
smoke_df <- smoke_df %>% dplyr::select(IOP2_ID, smoke) %>% mutate(IOP2_ID = as.factor(IOP2_ID))
```

Add to targets

```{r}
targets <- left_join(targets, smoke_df, by='IOP2_ID')
xtabs(~targets$smoke)
```

# Save Unfiltered Data

Attach `targets` rownames and remove PCs

```{r}
rownames(targets) <- targets$Basename
targets <- targets %>% 
  dplyr::select(-PC1, -PC2, -PC3, -PC4, -PC5)
```

Subset `betas` to include only target samples and check order

```{r}
betas <- betas[,colnames(betas) %in% rownames(targets)]
str(betas)
xtabs(~colnames(betas) == rownames(targets))
```

Subset annotation to include only probes in `betas`

```{r}
anno <- anno %>% filter(cpg %in% rownames(betas))
str(anno)

betas <- betas[rownames(betas) %in% anno$cpg,]
str(betas)
```

Save unfiltered values as a `SummarizedExperiment` object before probe removal

```{r}
methData_unfiltered <- SummarizedExperiment(
  assays=SimpleList(beta=betas), 
  colData=targets,
  rowData=anno)

methData_unfiltered

save(
  methData_unfiltered,
  file="../Processing/GOTO_methData-unfiltered.Rdata")

save(
  targets, 
  file="../Processing/GOTO_targets-unfiltered.Rdata")
```

Subset `RGset` and check column order

```{r}
RGset <- RGset[,colnames(RGset) %in% rownames(targets)]
RGset
xtabs(~colnames(RGset) == rownames(targets))
```

Save RGset

```{r}
RGset <- RGset[ , colnames(RGset) %in% colnames(betas)]

save(
  RGset, 
  file = "../Processing/GOTO_RGset-unfiltered.Rdata")
```

***

# Session Info

```{r}
sessionInfo()
```

Clear

```{r}
rm(list=ls())
```


